File: samples/java-replicatedentity-shopping-cart/src/main/java/com/example/shoppingcart/domain/ShoppingCart.java
/*
 * Copyright 2021 Lightbend Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.shoppingcart.domain;

import com.akkaserverless.javasdk.replicatedentity.ReplicatedCounterMap;
import com.akkaserverless.javasdk.replicatedentity.ReplicatedEntityContext;
import com.example.shoppingcart.ShoppingCartApi;
import com.google.protobuf.Empty;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class ShoppingCart extends AbstractShoppingCart {
  @SuppressWarnings("unused")
  private final String entityId;

  public ShoppingCart(ReplicatedEntityContext context) {
    this.entityId = context.entityId();
  }

  @Override
  public Effect<Empty> addItem(
      ReplicatedCounterMap<ShoppingCartDomain.Product> cart,
      ShoppingCartApi.AddLineItem addLineItem) {

    if (addLineItem.getQuantity() <= 0) {
      return effects().error("Cannot add negative quantity to item: " + addLineItem.getProductId());
    }

    ShoppingCartDomain.Product product =
        ShoppingCartDomain.Product.newBuilder()
            .setId(addLineItem.getProductId())
            .setName(addLineItem.getName())
            .build();

    cart.increment(product, addLineItem.getQuantity());
    return effects().update(cart).thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<Empty> removeItem(
      ReplicatedCounterMap<ShoppingCartDomain.Product> cart,
      ShoppingCartApi.RemoveLineItem removeLineItem) {

    ShoppingCartDomain.Product product =
        ShoppingCartDomain.Product.newBuilder()
            .setId(removeLineItem.getProductId())
            .setName(removeLineItem.getName())
            .build();

    if (!cart.containsKey(product)) {
      return effects().error("Item to remove is not in the cart: " + removeLineItem.getProductId());
    }

    cart.remove(product);
    return effects().update(cart).thenReply(Empty.getDefaultInstance());
  }

  @Override
  public Effect<ShoppingCartApi.Cart> getCart(
      ReplicatedCounterMap<ShoppingCartDomain.Product> cart,
      ShoppingCartApi.GetShoppingCart getShoppingCart) {

    List<ShoppingCartApi.LineItem> allItems =
        cart.keySet().stream()
            .map(
                product ->
                    ShoppingCartApi.LineItem.newBuilder()
                        .setProductId(product.getId())
                        .setName(product.getName())
                        .setQuantity(cart.get(product))
                        .build())
            .sorted(Comparator.comparing(ShoppingCartApi.LineItem::getProductId))
            .collect(Collectors.toList());

    return effects().reply(ShoppingCartApi.Cart.newBuilder().addAllItems(allItems).build());
  }

  @Override
  public Effect<Empty> removeCart(
      ReplicatedCounterMap<ShoppingCartDomain.Product> cart,
      ShoppingCartApi.RemoveShoppingCart removeShoppingCart) {

    return effects().delete().thenReply(Empty.getDefaultInstance());
  }
}

File: samples/scala-replicatedentity-shopping-cart/src/main/scala/com/example/shoppingcart/domain/ShoppingCart.scala
package com.example.shoppingcart.domain

import com.akkaserverless.scalasdk.replicatedentity.ReplicatedCounterMap
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntity
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntityContext
import com.example.shoppingcart
import com.google.protobuf.empty.Empty

// This class was initially generated based on the .proto definition by Akka Serverless tooling.
//
// As long as this file exists it will not be overwritten: you can maintain it yourself,
// or delete it so it is regenerated as needed.

/** A replicated entity. */
// tag::class[]
class ShoppingCart(context: ReplicatedEntityContext) extends AbstractShoppingCart { // <1>
  // end::class[]

  /** Command handler for "AddItem". */
  // tag::addItem[]
  def addItem(
      cart: ReplicatedCounterMap[Product],
      addLineItem: shoppingcart.AddLineItem): ReplicatedEntity.Effect[Empty] = {
    if (addLineItem.quantity <= 0) { // <1>
      effects.error(s"Quantity for item ${addLineItem.productId} must be greater than zero.")
    } else {
      val product = Product(addLineItem.productId, addLineItem.name) // <2>
      val updatedCart = cart.increment(product, addLineItem.quantity) // <3>

      effects
        .update(updatedCart) // <4>
        .thenReply(Empty.defaultInstance) // <5>
    }
  }
  // end::addItem[]

  /** Command handler for "RemoveItem". */
  def removeItem(
      cart: ReplicatedCounterMap[Product], // <1>
      removeLineItem: shoppingcart.RemoveLineItem): ReplicatedEntity.Effect[Empty] = {
    val product = Product(removeLineItem.productId, removeLineItem.name) // <2>

    if (!cart.contains(product)) {
      effects.error(s"Item to remove is not in the cart: ${removeLineItem.productId}")
    } else {
      effects
        .update(cart.remove(product))
        .thenReply(Empty.defaultInstance)
    }
  }

  /** Command handler for "GetCart". */
  // tag::getCart[]
  def getCart(
      cart: ReplicatedCounterMap[Product], // <1>
      getShoppingCart: shoppingcart.GetShoppingCart): ReplicatedEntity.Effect[shoppingcart.Cart] = {

    val allItems =
      cart.keySet
        .map { product =>
          val quantity = cart.get(product).getOrElse(0L)
          shoppingcart.LineItem(product.id, product.name, quantity)
        }
        .toSeq
        .sortBy(_.productId)

    val apiCart = shoppingcart.Cart(allItems) // <2>
    effects.reply(apiCart)
  }
  // end::getCart[]

  /** Command handler for "RemoveCart". */
  // tag::removeCart[]
  def removeCart(
      cart: ReplicatedCounterMap[Product],
      removeShoppingCart: shoppingcart.RemoveShoppingCart): ReplicatedEntity.Effect[Empty] =
    effects.delete // <1>
      .thenReply(Empty.defaultInstance)
  // end::removeCart[]

}
